Mathlib/RingTheory/PicardGroup.lean

```diff
@@ -0,0 +1,425 @@
+
+theorem toPic_dual : toPic R (Dual R M) = (toPic R M)⁻¹ :=
+  congr_arg (equivShrink _) <| Units.ext <| by
+    rw [toPic, Equiv.symm_apply_apply]
+    exact Quotient.sound ⟨(Finite.reprEquiv R _ ≪≫ₗ (Finite.reprEquiv R _).dualMap).toModuleIso⟩
+
+theorem _root_.CommRing.Pic.inv_eq_dual (M : Pic R) : M⁻¹ = toPic R (Dual R M) := by
+  rw [toPic_dual, toPic_eq_self]
+
+theorem _root_.CommRing.Pic.mul_eq_tensor (M N : Pic R) : M * N = toPic R (M ⊗[R] N) := by
+  rw [toPic_tensor, toPic_eq_self, toPic_eq_self]
+
+theorem subsingleton_Pic_iff : Subsingleton (Pic R) ↔
+    ∀ (M : Type u) [AddCommGroup M] [Module R M], Module.Invertible R M → Free R M := by
+  refine Iff.trans ?_ (forall₄_congr fun _ _ _ _ ↦ free_iff_linearEquiv.symm)
+  exact ⟨fun _ M _ _ _ ↦ toPic_eq_one_iff.mp (Subsingleton.elim ..), fun h ↦ ⟨fun M N ↦
+    Pic.ext_iff.mpr ⟨(h _ _ _ inferInstance).some ≪≫ₗ (h _ _ _ inferInstance).some.symm⟩⟩⟩
+
+instance [Subsingleton (Pic R)] : Free R M :=
+  have := subsingleton_Pic_iff.mp ‹_› (Finite.repr R M) inferInstance
+  .of_equiv (Finite.reprEquiv R M)
+
+/- TODO: it's still true that an invertible module over a (commutative) local semiring is free;
+  in fact invertible modules over a semiring is Zariski-locally free.
+  See [BorgerJun2024], Theorem 11.7. -/
+instance [IsLocalRing R] : Subsingleton (Pic R) :=
+  subsingleton_Pic_iff.mpr fun _ _ _ _ ↦ free_of_flat_of_isLocalRing
+
+end Module.Invertible
+
+end PicardGroup
+
+namespace Submodule
+
+open Module Invertible
+
+variable {R A : Type*} [CommSemiring R]
+
+section Semiring
+
+variable [Semiring A] [Algebra R A] [FaithfulSMul R A]
+
+open LinearMap in
+instance projective_unit (I : (Submodule R A)ˣ) : Projective R I := by
+  obtain ⟨T, T', hT, hT', one_mem⟩ := mem_span_mul_finite_of_mem_mul (I.inv_mul ▸ one_le.mp le_rfl)
+  classical
+  rw [← Set.image2_mul, ← Finset.coe_image₂, mem_span_finset] at one_mem
+  set S := T.image₂ (· * ·) T'
+  obtain ⟨r, hr⟩ := one_mem
+  choose a ha b hb eq using fun i : S ↦ Finset.mem_image₂.mp i.2
+  let f : I →ₗ[R] S → R := .pi fun i ↦ (LinearEquiv.ofInjective
+      (Algebra.linearMap R A) (FaithfulSMul.algebraMap_injective R A)).symm.comp <|
+    restrict (mulRight R (r i • a i)) fun x hx ↦ by
+      rw [← one_eq_range, ← I.mul_inv]; exact mul_mem_mul hx (I⁻¹.1.smul_mem _ <| hT <| ha i)
+  let g : (S → R) →ₗ[R] I := .lsum _ _ ℕ fun i ↦ .toSpanSingleton _ _ ⟨b i, hT' <| hb i⟩
+  refine .of_split f g (LinearMap.ext fun x ↦ Subtype.ext ?_)
+  simp only [f, g, lsum_apply, comp_apply, sum_apply, toSpanSingleton_apply, proj_apply, pi_apply]
+  simp_rw [restrict_apply, mulRight_apply, id_apply, coe_sum, coe_smul, Algebra.smul_def,
+    ← Algebra.coe_linearMap, LinearEquiv.coe_toLinearMap, LinearEquiv.ofInjective_symm_apply,
+    mul_assoc, Algebra.coe_linearMap, ← Algebra.smul_def, ← Finset.mul_sum, eq,
+    (Finset.sum_coe_sort ..).trans hr.2, mul_one]
+
+theorem projective_of_isUnit {I : Submodule R A} (hI : IsUnit I) : Projective R I :=
```

Comment:
I think if you want the dot notation to work you would want `I` to be an explicit variable.

Reply:
No longer necessary in Lean 4!